### ISSUE TITLE
Error: No access, refresh token or API key is set. setCredentials

### ISSUE THREAD

**Issue by begimai** [google/google-api-nodejs-client#894]
<your detailed question and code samples go here>
// Retrieve tokens via token exchange explained above or set them:
oauth2Client.setCredentials({
  access_token: 'ACCESS TOKEN HERE',
  refresh_token: 'REFRESH TOKEN HERE'
  // Optional, provide an expiry_date (milliseconds since the Unix Epoch)
  // expiry_date: (new Date()).getTime() + (1000 * 60 * 60 * 24 * 7)
});
   
this used to work, lately when I updated version, it says 
`TypeError: oauth2Client.setCredentials is not a function`

somewhere in the issues, it was suggested to switch to 
oauth2Client.setCredentials = {
  access_token: process.env['ACCESS_TOKEN'],
  refresh_token: process.env['REFRESH_TOKEN'],
  expiry_date: true
};

and now it returns Error: No access, refresh token or API key is set. 
How to solve this problem?

Thank you in advance!

**Comment by bantini** [comment]
@begimai Does this thread help?

https://github.com/google/google-api-nodejs-client/issues/869

**Comment by begimai** [comment]
Nope
 `oAuth2Client.credentials = credentials` does not assign credentials, everything is undefined

**Comment by bantini** [comment]
@begimai Can you share the code that you are using?

**Comment by begimai** [comment]
https://gist.github.com/begimai/36cb1239523bdea6b62dce1c7ebc4034

**Comment by piavgh** [comment]
@bantini : Hi, I got the same problem.
Please check it here https://github.com/google/google-api-nodejs-client/issues/869#issuecomment-352401977

Thank you.

**Comment by piavgh** [comment]
I tried to call Google API using request, and it works properly

```
GoogleTokens.findOne({isObsolete: false}).then((token) => {
                let subscriptionsGetUrl = "https://www.googleapis.com/androidpublisher/v2/applications/" + req.query.packageName + "/purchases/subscriptions/" + req.query.subscriptionId + "/tokens/" + req.query.token + "?access_token=" + token.accessToken;

                request(subscriptionsGetUrl, function (error, response, body) {
                    if (error) {
                        throw error;
                    }

                    res.success(JSON.parse(body).expiryTimeMillis);
                });
            });
```

**Comment by mike-seekwell** [comment]
@piavgh @bantini Here's what worked for me:

		var oauth2Client = new auth.OAuth2(clientId, clientSecret, redirectUrl);
		oauth2Client.credentials = {
		                refresh_token: 'your_refresh_token'
		};
		oauth2Client.refreshAccessToken(function(err, tokens){
		console.log(tokens)
		oauth2Client.credentials = {access_token : tokens.access_token}
		callback(oauth2Client);

**Comment by begimai** [comment]
Thank you, everyone
I will try out
My solution was to use "googleapis": "^22.2.0"

**Comment by zoellner** [comment]
Having the same issue with v23.0.0
This should be marked as breaking change. It seems to be due to updates to google/google-auth-library-nodejs

**Comment by PabloValor** [comment]
@begimai which google-auth-library version did you use it?

**Comment by begimai** [comment]
@PabloValor 
This one google.auth.OAuth2

**Comment by JustinBeckwith** [comment]
Greetings folks!  This has been fixed in version [25.0.0](https://github.com/google/google-api-nodejs-client/releases/tag/v25.0.0).  Please do let us know if you're still running into issues!

**Comment by opyate** [comment]
FWIW, here's the modified (from [this](https://github.com/google/google-api-nodejs-client/blob/master/samples/oauth2.js)) code I use on Cloud Functions which is limited to Node 6.14.0

```
const {google} = require('googleapis')
const plus = google.plus('v1')
const oauth2Client = new google.auth.OAuth2('your-client-id', 'secret', 'redirect')
google.options({ auth: oauth2Client })
const fromAuthCode = (authCode, userId) => {
  return oauth2Client.getToken(authCode).then(tokenResponse => {
    oauth2Client.setCredentials(tokenResponse.tokens)
    return plus.people.get({ userId: userId }).then(response => {
      return response
    })
  })
}
```
